### PR TITLE
[SPARK-20124][SQL] Join reorder should keep the same order of final project attributes

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

Join reorder algorithm should keep exactly the same order of output attributes in the top project. 
For example, if user want to select a, b, c, after reordering, we should output a, b, c in the same order as specified by user, instead of b, a, c or other orders.

## How was this patch tested?

A new test case is added in `JoinReorderSuite`.

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @cloud-fan please review

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala [hunk 1]
```diff
@@ -127,7 +127,8 @@ abstract class PlanTest extends SparkFunSuite with PredicateHelper {
         (sameJoinPlan(j1.left, j2.left) && sameJoinPlan(j1.right, j2.right)) ||
           (sameJoinPlan(j1.left, j2.right) && sameJoinPlan(j1.right, j2.left))
       case _ if plan1.children.nonEmpty && plan2.children.nonEmpty =>
-        (plan1.children, plan2.children).zipped.forall { case (c1, c2) => sameJoinPlan(c1, c2) }
+        plan1 == plan2 &&
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: Why should they be equal? This implies that the entire subree is the same as well right?

Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: It compares the two plans themselves, e.g. I also want to make sure projectLists are equal in Project nodes. Subtree comparison is done by recursively calling `sameJoinPlan` for their child/children.

Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: Yes you are right. There's some confusion here.

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Pull request closed without merging**